cmd/git_utils.go

```diff
@@ -0,0 +1,215 @@
+}
+
+const trimChars = "' \r\n"
+
+func stringBefore(value string, searchValue string) string {
+	// Get substring before a string.
+
+	gitURLElements := strings.Split(value, searchValue)
+	if len(gitURLElements) == 0 {
+		return ""
+	}
+	return gitURLElements[0]
+
+}
+
+func stringAfter(value string, searchValue string) string {
+	// Get substring after a string.
+	position := strings.LastIndex(value, searchValue)
+	if position == -1 {
+		return ""
+	}
+	adjustedPosition := position + len(searchValue)
+	if adjustedPosition >= len(value) {
+		return ""
+	}
+	return value[adjustedPosition:]
+}
+func stringBetween(value string, pre string, post string) string {
+	// Get substring between two strings.
+	positionBegin := strings.Index(value, pre)
+	if positionBegin == -1 {
+		return ""
+	}
+	positionEnd := strings.Index(value, post)
+	if positionEnd == -1 {
+		return ""
+	}
+	positionBeginAdjusted := positionBegin + len(pre)
+	if positionBeginAdjusted >= positionEnd {
+		return ""
+	}
+	return value[positionBeginAdjusted:positionEnd]
+}
+
+//RunGitFindBranc issues git status
+func GetGitInfo(dryrun bool) (GitInfo, error) {
+	var gitInfo GitInfo
+	version, vErr := RunGitVersion(false)
+	if vErr != nil {
+		return gitInfo, vErr
+	}
+	if version == "" {
+		return gitInfo, errors.Errorf("git does not appear to be available")
+	}
+
+	Debug.log("git version: ", version)
+
+	kargs := []string{"status", "-sb"}
+
+	output, gitErr := RunGit(kargs, dryrun)
+	if gitErr != nil {
+		return gitInfo, gitErr
+	}
+
+	lineSeparator := "\n"
+	if runtime.GOOS == "windows" {
+		lineSeparator = "\r\n"
+	}
+	outputLines := strings.Split(output, lineSeparator)
+
+	var changeLines []string
+	const noCommits = "## No commits yet on "
+	const branchPrefix = "## "
+	const branchSeparatorString = "..."
+	for _, value := range outputLines {
+
+		branchFound := false
+		if strings.Trim(value, trimChars) != "" {
+
+			if strings.HasPrefix(value, branchPrefix) {
+				gitInfo.Branch = strings.Trim(stringBetween(value, branchPrefix, branchSeparatorString), trimChars)
+				gitInfo.Upstream = strings.Trim(stringAfter(value, branchSeparatorString), trimChars)
+				branchFound = true
+			}
+			if strings.Contains(value, noCommits) {
+
+				branchFound = true
+				gitInfo.Branch = stringAfter(value, noCommits)
+
+			}
+			// this must be a change line
+			if !branchFound {
+				changeLines = append(changeLines, value)
+			}
+		}
+
+	}
+	gitInfo.Changes = append(gitInfo.Changes, changeLines...)
+	if gitInfo.Upstream != "" {
+		gitInfo.RemoteURL, gitErr = RunGitConfigLocalRemoteOriginURL(gitInfo.Upstream, dryrun)
+		if gitErr != nil {
+			Info.Logf("Could not construct repository URL %v", gitErr)
+		}
+
+	} else {
+		Info.log("Unable to determine origin to compute repository URL")
+	}
+
+	gitInfo.Commit, gitErr = RunGitGetLastCommit(gitInfo.RemoteURL, dryrun)
+	if gitErr != nil {
+		Info.log("Received error getting current commit: ", gitErr)
+		//return gitInfo, gitErr
```

Comment:
Should we return an error, or return the `gitInfo` struct without the current commit? I am not sure

Reply:
The thought here was not to error out, if we can't find the commit.